app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php

```diff
@@ -0,0 +1,154 @@
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\Api\SearchCriteriaBuilderFactory;
+use Magento\InventoryReservations\Model\ReservationBuilderInterface;
+use Magento\InventoryReservationsApi\Api\AppendReservationsInterface;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryApi\Api\SourceItemsSaveInterface;
+use Magento\InventorySales\Model\StockByWebsiteIdResolver;
+use Magento\Framework\Exception\LocalizedException;
+
+class SourceDeductionProcessor implements ObserverInterface
+{
+    /**
+     * @var SearchCriteriaBuilderFactory
+     */
+    private $searchCriteriaBuilderFactory;
+
+    /**
+     * @var SourceItemRepositoryInterface
+     */
+    private $sourceItemRepository;
+
+    /**
+     * @var ReservationBuilderInterface
+     */
+    private $reservationBuilder;
+
+    /**
+     * @var AppendReservationsInterface
+     */
+    private $appendReservations;
+
+    /**
+     * @var SourceItemsSaveInterface
+     */
+    private $sourceItemsSave;
+
+    /**
+     * @var StockByWebsiteIdResolver
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * SourceDeductionProcessor constructor.
+     * @param SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory
+     * @param SourceItemRepositoryInterface $sourceItemRepository
+     * @param ReservationBuilderInterface $reservationBuilder
+     * @param AppendReservationsInterface $appendReservations
+     * @param SourceItemsSaveInterface $sourceItemsSave
+     * @param StockByWebsiteIdResolver $stockByWebsiteIdResolver
+     */
+    public function __construct(
+        SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory,
+        SourceItemRepositoryInterface $sourceItemRepository,
+        ReservationBuilderInterface $reservationBuilder,
+        AppendReservationsInterface $appendReservations,
+        SourceItemsSaveInterface $sourceItemsSave,
+        StockByWebsiteIdResolver $stockByWebsiteIdResolver
+    ) {
+        $this->searchCriteriaBuilderFactory = $searchCriteriaBuilderFactory;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->reservationBuilder = $reservationBuilder;
+        $this->appendReservations = $appendReservations;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Shipment $shipment */
+        $shipment = $observer->getEvent()->getShipment();
+        if ($shipment->getOrigData('entity_id')) {
+            return $this;
+        }
+
+        $order = $shipment->getOrder();
+
+        // I'm not sure about websiteId
+        $websiteId = $order->getStore()->getWebsiteId();
+
+        foreach ($shipment->getItems() as $item) {
+            if (!$item->getSources()) {
```

Comment:
there are no get/setSource methods in the 
`\Magento\Sales\Api\Data\ShipmentItemInterface`